File: crypto_screener/factor_explanations.py
from __future__ import annotations

from typing import Any

from .factor_definitions import DIRECTIONAL_FACTORS
from .scoring import to_float


def reason_for(row: dict[str, Any], side: str) -> str:
    parts: list[str] = []
    factors = row.get("factors", {})
    scores = row.get("scores", {})
    quality_flags = row.get("data_quality_flags") or []

    _append_metric(parts, "24h", row.get("price_change_24h_pct"), "{:+.2f}%")
    _append_metric(parts, "OI", row.get("oi_change_24h_pct"), "{:+.2f}%")
    _append_metric(parts, "funding", row.get("funding_rate_pct"), "{:+.4f}%")
    if row.get("long_short_ratio") is not None:
        parts.append(f"L/S {float(row['long_short_ratio']):.2f}")
    if scores.get("factor_score") is not None:
        parts.append(f"factor {scores['factor_score']:+.2f}")
    if scores.get("confidence_score") is not None:
        parts.append(f"confidence {scores['confidence_score']:.0f}")
    if row.get("signal_conflict_label") and row.get("signal_conflict_label") != "aligned":
        parts.append(f"signals {row['signal_conflict_label']}")
    if row.get("technical_setup"):
        parts.append(f"tech {row['technical_setup']}")

    strongest = sorted(
        (
            (name, numeric_value)
            for name, value in factors.items()
            if name in DIRECTIONAL_FACTORS and (numeric_value := to_float(value)) is not None
        ),
        key=lambda item: abs(item[1]),
        reverse=True,
    )[:2]
    for name, value in strongest:
        if abs(value) >= 0.5:
            parts.append(f"{name} {value:+.2f}")

    if side == "fade-long":
        parts.append("crowded long conditions")
    if side == "squeeze-risk":
        parts.append("crowded short conditions")
    if quality_flags:
        parts.append("excluded: " + ", ".join(str(flag) for flag in quality_flags))
    return "; ".join(parts)


def _append_metric(parts: list[str], label: str, value: Any, fmt: str) -> None:
    numeric = to_float(value)
    if numeric is not None:
        parts.append(f"{label} {fmt.format(numeric)}")

File: apps/api/src/pipeline/factorExplanations.ts
import { DIRECTIONAL_FACTORS } from './factorDefinitions.js';
import { toFloat } from './scoring.js';
import type { Row } from './types.js';
import { asRecord } from './types.js';

// Sign comes from the pre-rounding value (handles -0), not the rounded value.
export function formatSigned(value: number, decimals: number): string {
  const sign = value < 0 || Object.is(value, -0) ? '-' : '+';
  return `${sign}${Math.abs(value).toFixed(decimals)}`;
}

function appendMetric(
  parts: string[],
  label: string,
  value: unknown,
  decimals: number,
  suffix: string,
): void {
  const numeric = toFloat(value);
  if (numeric !== null) {
    parts.push(`${label} ${formatSigned(numeric, decimals)}${suffix}`);
  }
}

export function reasonFor(row: Row, side: string): string {
  const parts: string[] = [];
  const factors = asRecord(row.factors);
  const scores = asRecord(row.scores);
  const qualityFlags = Array.isArray(row.data_quality_flags) ? row.data_quality_flags : [];

  appendMetric(parts, '24h', row.price_change_24h_pct, 2, '%');
  appendMetric(parts, 'OI', row.oi_change_24h_pct, 2, '%');
  appendMetric(parts, 'funding', row.funding_rate_pct, 4, '%');
  if (row.long_short_ratio !== null && row.long_short_ratio !== undefined) {
    const ratio = toFloat(row.long_short_ratio) ?? 0.0;
    parts.push(`L/S ${ratio.toFixed(2)}`);
  }
  const factorScore = toFloat(scores.factor_score);
  if (factorScore !== null) {
    parts.push(`factor ${formatSigned(factorScore, 2)}`);
  }
  const confidenceScore = toFloat(scores.confidence_score);
  if (confidenceScore !== null) {
    parts.push(`confidence ${confidenceScore.toFixed(0)}`);
  }
  const conflictLabel = row.signal_conflict_label;
  if (conflictLabel && conflictLabel !== 'aligned') {
    parts.push(`signals ${String(conflictLabel)}`);
  }
  if (row.technical_setup) {
    parts.push(`tech ${String(row.technical_setup)}`);
  }

  const strongest = Object.entries(factors)
    .filter((entry): entry is [string, unknown] => DIRECTIONAL_FACTORS.includes(entry[0]))
    .map(([name, value]) => [name, toFloat(value)] as const)
    .filter((entry): entry is [string, number] => entry[1] !== null)
    .sort((a, b) => Math.abs(b[1]) - Math.abs(a[1]))
    .slice(0, 2);
  for (const [name, value] of strongest) {
    if (Math.abs(value) >= 0.5) {
      parts.push(`${name} ${formatSigned(value, 2)}`);
    }
  }

  if (side === 'fade-long') {
    parts.push('crowded long conditions');
  }
  if (side === 'squeeze-risk') {
    parts.push('crowded short conditions');
  }
  if (qualityFlags.length > 0) {
    parts.push(`excluded: ${qualityFlags.map((flag) => String(flag)).join(', ')}`);
  }
  return parts.join('; ');
}
